src/EditorFeatures/Core/Implementation/Structure/RoslynOutliningRegionTag.cs

```diff
@@ -0,0 +1,148 @@
+// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Windows.Media;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Structure;
+using Microsoft.CodeAnalysis.Text.Shared.Extensions;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Text.Projection;
+using Microsoft.VisualStudio.Text.Tagging;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.Structure
+{
+    // Our implementation of an outlining region tag.  The collapsedHintForm
+    // is dynamically created using an elision buffer over the actual text
+    // we are collapsing.
+    internal class RoslynOutliningRegionTag : IOutliningRegionTag
+    {
+        private const string OutliningRegionTextViewRole = nameof(OutliningRegionTextViewRole);
+
+        private const string Ellipsis = "...";
+        private const int MaxPreviewText = 1000;
+
+        private readonly ITextEditorFactoryService _textEditorFactoryService;
+        private readonly IProjectionBufferFactoryService _projectionBufferFactoryService;
+        private readonly IEditorOptionsFactoryService _editorOptionsFactoryService;
+
+        private readonly ITextBuffer _subjectBuffer;
+        private readonly ITrackingSpan _hintSpan;
+
+        public bool IsDefaultCollapsed => BlockSpan.IsDefaultCollapsed;
+        public bool IsImplementation => BlockSpan.AutoCollapse;
+        public object CollapsedForm => BlockSpan.BannerText;
+
+        protected readonly BlockSpan BlockSpan;
+
+        public RoslynOutliningRegionTag(
+            ITextEditorFactoryService textEditorFactoryService,
+            IProjectionBufferFactoryService projectionBufferFactoryService,
+            IEditorOptionsFactoryService editorOptionsFactoryService,
+            ITextSnapshot snapshot,
+            BlockSpan outliningSpan)
+        {
+            _textEditorFactoryService = textEditorFactoryService;
+            _projectionBufferFactoryService = projectionBufferFactoryService;
+            _editorOptionsFactoryService = editorOptionsFactoryService;
+            _subjectBuffer = snapshot.TextBuffer;
+            BlockSpan = outliningSpan;
+
+            _hintSpan = snapshot.CreateTrackingSpan(BlockSpan.HintSpan.ToSpan(), SpanTrackingMode.EdgeExclusive);
+        }
+
+        public object CollapsedHintForm =>
+            new ViewHostingControl(CreateElisionBufferView, CreateElisionBuffer);
```

Comment:
Kinda hate that this expensive allocation is a property getter. I suppose its defined that way on IOutliningRegionTag?